re-enable some pagestore testing

grandinj opened the pull request:
at some point I managed to remove the pagestore testing, put it back

grandinj committed:
re-enable some pagestore testing

at some point I managed to remove the pagestore testing, put it back

grandinj committed:
exclude joins.sql from pagestore test mode

grandinj committed:
make Session::toString() more robust again null

for debugging purposes

grandinj committed:
I didn't mean to test PageStore in mvcc mode

that is experimental

katzyn commented on h2/src/test/org/h2/test/scripts/TestScript.java:
```diff
@@ -88,7 +88,10 @@ public void test() throws Exception {
         testScript("derived-column-names.sql");
         testScript("information_schema.sql");
-        testScript("joins.sql");
+        if (!config.mvStore) {
```
This condition should be reversed.

grandinj committed:
we don't report output line no anymore

we report input line no, which is more useful

grandinj committed:
got the condition the wrong way around here

katzyn commented:
Just for information:

PageStore indexes with `ASC NULLS LAST` and `DESC NULLS FIRST` are broken, looks like `NULLS *` is ignored by them. This is the reason of failures in `median.sql`. We don't have other tests for this, but I've just checked that regular queries with `ORDER BY` with such indexes are also broken. If `index sorted` feature is available it leads to wrong sorting that is different from explicit sorting without an index. MVStore is not affected, its sorting is compatible with regular sorting without an index.

Probably we need some additional tests for indexes to detect such failures in more oblivious places.

katzyn commented:
I sent #1085 with simple pure SQL tests for this problem.

grandinj commented:
@katzyn thanks, feel free to fix any of the PageStore issues that take your fancy, I'm going to work on this on and off slowly

grandinj committed:
support custom datatypes in PageStore engin

grandinj committed:
don't need to test pageStore && multithread

grandinj committed:
only run mvcc tests in mvcc mode

grandinj committed:
this isn't throwing anymore

not sure why

grandinj committed:
fix choice of database file suffix

grandinj committed:
explicitly disable mvstore when not using it

otherwise we use it for tests where we actually want pagestore

grandinj committed:
catching NullPointerException is very bad practice

and makes finding real NPE problems super hard

grandinj committed:
cannot insert null SpatialKey values into the map

fixes TestSpatial in pagestore mode

grandinj committed:
Merge remote-tracking branch 'upstream/master' into testing_pagestore

grandinj commented:
perhaps @katzyn  could look at the Date PageStore failure? You seem to be an expert in date stuff :-)

11:44:18 00:05.679 org.h2.test.db.TestDateStorage Expected: 2000-01-01 15<*>:00:00 (19) actual: 2000-01-01 16:00:00 (19) Africa/Juba>sun.util.calendar.ZoneInfo[id="Africa/Abidjan",offset=0,dstSavings=0,useDaylight=false,transitions=3,lastRule=null]
ERROR: FAIL java.lang.AssertionError: Expected: 2000-01-01 15<*>:00:00 (19) actual: 2000-01-01 16:00:00 (19) Africa/Juba>sun.util.calendar.ZoneInfo[id="Africa/Abidjan",offset=0,dstSavings=0,useDaylight=false,transitions=3,lastRule=null] java.lang.AssertionError: Expected: 2000-01-01 15<*>:00:00 (19) actual: 2000-01-01 16:00:00 (19) Africa/Juba>sun.util.calendar.ZoneInfo[id="Africa/Abidjan",offset=0,dstSavings=0,useDaylight=false,transitions=3,lastRule=null] ------------------------------
java.lang.AssertionError: Expected: 2000-01-01 15<*>:00:00 (19) actual: 2000-01-01 16:00:00 (19) Africa/Juba>sun.util.calendar.ZoneInfo[id="Africa/Abidjan",offset=0,dstSavings=0,useDaylight=false,transitions=3,lastRule=null]
	at org.h2.test.TestBase.fail(TestBase.java:472)
	at org.h2.test.TestBase.assertEquals(TestBase.java:815)
	at org.h2.test.db.TestDateStorage.testMoveDatabaseToAnotherTimezone(TestDateStorage.java:204)
	at org.h2.test.db.TestDateStorage.test(TestDateStorage.java:44)
	at org.h2.test.TestBase.runTest(TestBase.java:142)
	at org.h2.test.TestAll.addTest(TestAll.java:998)
	at org.h2.test.TestAll.test(TestAll.java:871)
	at org.h2.test.TestAll.runTests(TestAll.java:650)
	at org.h2.test.TestAll.testAll(TestAll.java:562)
	at org.h2.test.TestAll.run(TestAll.java:511)
	at org.h2.test.TestAll.main(TestAll.java:451)

katzyn commented:
OK, I'll take a look on it.

katzyn commented:
This test is currently executed only in PageStore mode (probably because there is no reason to test it with MVStore that stores date-time values in local time).

There is a hard-coded configuration option for PageStore: 
https://github.com/h2database/h2database/blob/0468cb75cd12d6bc8b206f738234ca88f7b982d7/h2/src/main/org/h2/store/Data.java#L96-L103
This test completes normally if this constant is changed to `true`. Probably this test was designed for this mode.

But with `STORE_LOCAL_TIME = false` this test is useless. On Java 10 only 57% of conversions tested by this test are successful, the remaining 43% have different offsets due to differences in DST transitions between timezones.

I think that we should disable this test in all modes for now.

MVStore is much slower on this test, BTW. For example, on my system `getConnection()` calls on insertion of values in `testMoveDatabaseToAnotherTimezone()` take about 4 seconds in PageStore mode and about 25 seconds on MVStore mode.

grandinj commented:
Thanks for checking that

> MVStore is much slower on this test, BTW. For 

I guess we still have some optimisation to do :-)

grandinj committed:
disable testMoveDatabaseToAnotherTimezone

grandinj committed:
only run mvcc test in mvcc mode

fixes exception in pagestore mode

grandinj committed:
don't compact if store already closed

fixes exception in TestSpatial in pagestore mode

katzyn commented:
I merged a fix for PageStore indexes with `ASC NULLS LAST` and `DESC NULLS FIRST`.

grandinj commented:
@katzyn Nice, thanks! Got a few more fixes on machine too, so we're getting closer to having pagestore testing nicely again

grandinj committed:
Merge remote-tracking branch 'upstream/master' into testing_pagestore

katzyn commented:
I also merged a fix for `ON DUPLICATE KEY UPDATE` that resolves test failures in `TestDuplicateKeyUpdate` and `TestCompatibility`.

There is a significant problem with two-phase commit in PageStore. (MVStore also had a big problem, but I resolved it some time ago.)

`PREPARE TRANSACTION name` + `commit()` / `rollback()` + `SHUTDOWN [IMMEDIATELY]` sequence leads to in-doubt transaction in PageStore. But this is not expected, because successful `commit()` or `rollback()` calls with 2PC should ensure that transaction is either committed or rolled back and discarded after that.

katzyn commented:
`PREPARE TRANSACTION name` + `commit()` + `SHUTDOWN IMMEDIATELY` also leads to loss of committed data in PageStore mode.

katzyn commented on h2/src/test/org/h2/test/db/TestTriggersConstraints.java:
```diff
@@ -119,8 +118,7 @@ private void testDeleteInTrigger() throws SQLException {
         // this threw a NullPointerException
-        assertThrows(ErrorCode.ROW_NOT_FOUND_WHEN_DELETING_1, stat).
-                execute("update test set id = 2");
+        stat.execute("update test set id = 2");
```
Leftover comment above?

katzyn commented on h2/src/test/org/h2/test/scripts/TestScript.java:
```diff
@@ -90,7 +89,10 @@ public void test() throws Exception {
+        if (config.mvStore) {
+            // we get slightly different explain plan stuff here in PageStore mode
+            testScript("joins.sql");
+        }
```
I pushed a change that allows to set different expectations in different modes and fixed `joins.sql` with that functionality. So this test should not be disabled, it completes normally now in both MVStore and PageStore modes.

grandinj committed:
Merge remote-tracking branch 'upstream/master' into testing_pagestore

grandinj committed:
fix pagestore index calculations

which results in (as far as I can see) quite a few cases where we now pick the faster option of an index instead of scanning over the pagestore table data

katzyn commented:
There is an issue in `MergeUsing` too. I'm trying to debug it now.

grandinj commented:
@katzyn please check my commit 362c139 and see if you agree with my analysis.

Thanks for the help, the new TestScript functionality was particularly useful.

grandinj committed:
address some review comments

katzyn commented:
I can't definitely say if increased cost of `PageDataIndex` helps or not. Even in large tests deviation of execution time in different runs is too high.

katzyn commented:
Was this change made to resolve issues with TestOptimizations? Or there were other reasons?

It created a lot of differences with MVStore in `testScript.sql` and also in `joins.sql` that should be adjusted once again. Before this change only one test was different in `joins.sql`, but after it there are a few additional differences.

grandinj commented:
It was done to fix a bug in testMinimalCoveringIndexPlan() in TestCases, where it was picking the pagedata table i.e. the scan index rather than a minimal covering index.

It helps to understand that MVStore and PageStore have different approaches to storing the main row data. PageStore always has what it calls a "scan index" or PageDataIndex, which is effectively a table indexed by row id. In the PageStore world, even a primary key index is a separate index.

However, MVStore is different - if there is a primary key, the main row data is stored in a MVMap indexed by the primary key.

Hence all the differences in explain plan output. I don't think there is any reasonable way to get around this.

grandinj committed:
no need for this init() method

since the test() method already ignores anything except pagestore

grandinj committed:
fix TestLinkedTable test in pagestore mode

grandinj committed:
fix joins.sql in pagestore mode

grandinj committed:
disable the two last failing tests in pagestore mode

running out of enthusiam for debugging this, lets just get it into master so we can prevent even more damage

grandinj committed:
Merge remote-tracking branch 'upstream/master' into testing_pagestore

grandinj committed:
make pagestore testing mode more obvious in the test log

katzyn commented on h2/src/test/org/h2/test/synth/TestConcurrentUpdate.java:
```diff
@@ -37,8 +36,11 @@ public static void main(String... a) throws Exception {
+        }
         deleteDb("concurrent");
-        final String url = getURL("concurrent;MULTI_THREADED=TRUE", true);
+        final String url = getURL("concurrent", true);
```
BTW, this test looks a bit strange. `r.nextInt(3)` may return only 0, 1, or 2:
https://github.com/h2database/h2database/blob/5365e664b6560d757738919a7b5768e85e750890/h2/src/test/org/h2/test/synth/TestConcurrentUpdate.java#L66
But there is an additional unused block:
https://github.com/h2database/h2database/blob/5365e664b6560d757738919a7b5768e85e750890/h2/src/test/org/h2/test/synth/TestConcurrentUpdate.java#L81-L87

grandinj committed:
Merge remote-tracking branch 'upstream/master' into testing_pagestore

grandinj committed:
disable some more tests in pagestore mode

grandinj committed:
fix bug in SimpleResultSet::beforeFirst()

needs to throw a SQLException not a DbException

grandinj committed:
exclude more pagestore mode test

katzyn commented on h2/src/test/org/h2/test/unit/TestOldVersion.java:
```diff
@@ -114,6 +113,10 @@ private void checkResult(ResultSet rs) throws SQLException {
     private void testOldClientNewServer() throws Exception {
+        // TODO fails in pagestore mode
+        if (!config.mvStore) {
+            return;
```
The whole test is PageStore-only. It should not be disabled, especially in such way.

I sent a possible fix in #1123.

katzyn replied:
I merged a fix. `TestOldVersion` now completes normally in PageStore mode on Java 8 and Java 9. I didn't test Java 7, however.

So this test can be enabled again.

grandinj committed:
Merge remote-tracking branch 'upstream/master' into testing_pagestore

grandinj committed:
re-enable this test

now that katzyn has fixed the underlying issue

grandinj committed:
fix another TestCrashAPI failure

grandinj committed:
disable another pagestore LOB test

grandinj committed:
fix TestMultiThreadedKernel in pagestore mode

by essentially copying over various fixes we made, from the MVStore class, to the RegularTable class

katzyn commented:
I reviewed the summary diff once again. I have no complains.

Perhaps we should fill some separate issues for all temporary disabled tests, otherwise we can forget about some existing problems that were found. We can create a PageStore label to mark them with it.

I noticed `TestTwoPhaseCommit.testInDoubtAfterShutdown()` that seems to be a regression, `TestMergeUsing` (probably never worked in PageStore mode), some problems with LOBs and geometries. May be I missed something.
